File: macos/hd_sentry/Sources/hd_sentry/HdSentryCrashHandler.swift
import Foundation

enum HdSentryCrashHandler {
  private static var installed = false
  private static var previousExceptionHandler: (@convention(c) (NSException) -> Void)?

  private static var crashReportWritten: Int32 = 0

  private static let platform = "macos"

  static func install() {
    if installed { return }
    installed = true
    HdSentryCrashStore.configure()

    previousExceptionHandler = NSGetUncaughtExceptionHandler()
    NSSetUncaughtExceptionHandler(exceptionHandler)

    installSignalHandlers()
  }

  private static func writeReportOnce(
    type: String,
    message: String,
    stackTrace: String,
    threadName: String? = nil
  ) {
    guard OSAtomicCompareAndSwap32Barrier(0, 1, &crashReportWritten) else {
      return
    }
    try? HdSentryCrashStore.writeReport(
      platform: platform,
      type: type,
      message: message,
      stackTrace: stackTrace,
      threadName: threadName
    )
  }

  private static let exceptionHandler: @convention(c) (NSException) -> Void = { exception in
    let stack = exception.callStackSymbols.joined(separator: "\n")
    writeReportOnce(
      type: "uncaught_exception",
      message: exception.reason ?? exception.name.rawValue,
      stackTrace: stack,
      threadName: Thread.current.name
    )
    if let previous = previousExceptionHandler {
      previous(exception)
    }
  }

  private static func installSignalHandlers() {
    let signals: [Int32] = [SIGABRT, SIGBUS, SIGFPE, SIGILL, SIGSEGV, SIGTRAP]
    for sig in signals {
      Darwin.signal(sig, signalHandler)
    }
  }

  private static let signalHandler: @convention(c) (Int32) -> Void = { sig in
    Darwin.signal(sig, SIG_DFL)

    writeReportOnce(
      type: "signal",
      message: "Signal \(sig) received",
      stackTrace: Thread.callStackSymbols.joined(separator: "\n")
    )

    Darwin.raise(sig)
  }
}

File: linux/hd_sentry_crash_handler.c
#include "hd_sentry_crash_handler.h"

#include "hd_sentry_crash_store.h"
#include "hd_sentry_linux_stack_trace.h"

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>

static volatile sig_atomic_t g_crash_report_written = 0;
static gboolean g_installed = FALSE;

static void signal_handler(int sig) {
  signal(sig, SIG_DFL);
  if (g_crash_report_written == 0) {
    g_crash_report_written = 1;
    g_autofree gchar* msg =
        g_strdup_printf("Signal %d received", sig);
    g_autofree gchar* stack =
        hd_sentry_linux_stack_trace_capture_skip(2);
    hd_sentry_crash_store_write_report("linux", "signal", msg, stack);
  }
  raise(sig);
}

void hd_sentry_crash_handler_install_signals(void) {
  if (g_installed) {
    return;
  }
  g_installed = TRUE;

  hd_sentry_crash_store_configure();

  const int signals[] = {SIGABRT, SIGBUS, SIGFPE, SIGILL, SIGSEGV, SIGTRAP};
  for (gsize i = 0; i < G_N_ELEMENTS(signals); ++i) {
    signal(signals[i], signal_handler);
  }
}
